### PR TITLE
Y2K

### PR DESCRIPTION
**Pull request by cfry-propeller**
Changes to bambooR to support "Y2K". get_pto() and get_bench() had to have changes made (end_date and year parameters, respectively) in order to support our year-to-date angle.

### CODE REVIEW

File: R/get_pto.r [hunk 1]
```diff
@@ -4,59 +4,118 @@
...
+#' @param year int; year for which PTO records are desired; defaults to current year
+#' @param verbose logical; indicates if detailed output from httr calls should be provided; default FALSE
 #' @return tbl_df
 #'
 #' @examples
-#'
+#' \dontrun{
 #' user <- 'your_api_user'
 #' password <- 'your_password'
 #' employees <- get_pto(user=user,password=password)
```
Comment (adamgrimes) [inline R/get_pto.r]:
```suggestion
+#' startDate = lubridate::ymd("20190101")
+#' endDate = lubridate::ymd("20191231")
+#' employees <- get_pto(user = user, password = password, startDate = startDate, endDate = endDate)
```

File: R/get_pto.r [hunk 2]
```diff
@@ -4,59 +4,118 @@
...
-      purrr::map(.,'dates') %>%
-      purrr::map(.,names) %>%
-      purrr::map2_df(.x=.,.y=time_off %>% purrr::map(., `[`, c('id','employeeId')), function(x=.x,y=.y) unlist(x) %>% dplyr::as_data_frame(.) %>% dplyr::mutate(id=unlist(y)[1],employeeId=unlist(y)[2])) %>% dplyr::rename('Date'='value'),
+      purrr::map(., 'dates') %>%
+      purrr::map(., names) %>%
+      purrr::map2_df(
+        .x = .,
+        .y = time_off %>% purrr::map(., `[`, c('id', 'employeeId')),
+        function(x = .x, y = .y) unlist(x) %>%
+          dplyr::as_data_frame(.) %>%
```
Comment (adamgrimes) [inline R/get_pto.r]:
```suggestion
          tibble::enframe('id') %>%
```

File: R/get_pto.r [hunk 3]
```diff
@@ -4,59 +4,118 @@
...
-                                 purrr::map(.,'amount') %>%
-                                 purrr::map_df(., `[`, c('unit','amount')))
-  time_off <- dplyr::inner_join(time_off,time_off_days) %>%
-    dplyr::rename('PTO_ID'='id','Spent_Date'='Date','PTO_updatedDate'='lastChanged','Employee_bambooID'='employeeId','PTO_Status'='status','PTO_Type'='name','PTO_timesUnit'='unit','PTO_Time'='amount') %>%
-    dplyr::select(PTO_ID,Employee_bambooID,Spent_Date,PTO_updatedDate,PTO_Status,PTO_Type,PTO_timesUnit,PTO_Time,Hours) %>%
-    dplyr::mutate_at(dplyr::vars('PTO_ID','Employee_bambooID'),dplyr::funs(as.integer(.))) %>%
-    dplyr::mutate_at(dplyr::vars(colnames(df)[stringr::str_detect(names(df),'Date')]),dplyr::funs(lubridate::ymd(.))) %>%
-    dplyr::mutate_at(dplyr::vars('PTO_Time','Hours'),dplyr::funs(as.numeric(.)))
+      purrr::map_df(., function(x) unlist(x) %>%
+                        dplyr::as_data_frame(.)) %>%
```
Comment (adamgrimes) [inline R/get_pto.r]:
```suggestion
                        tibble::enframe('dates')) %>%
```

File: R/get_pto.r (hunk 2 above)
Comment (adamgrimes) [reply R/get_pto.r]: Truly don't know if that is right

**Review by adamgrimes — COMMENTED** [review]
I didn't review the man changes. The enframe changes are very low priority, but I can comparedf data sets before and after changes and it creates the same data

File: R/get_bench.r [hunk 4]
```diff
@@ -4,40 +4,85 @@
...
+  # specified
+  # The only scenario the below doesn't cover is a situation where the bench
+  # record covers the entire year, which seems unrealistic
+  if(!is.null(year)){
+    df <-
+      dplyr::filter(
+        df,
+        lubridate::year(Bench_startDate) == year |
+          lubridate::year(Bench_endDate) == year |
+          is.na(Bench_endDate))
```
Comment (adamgrimes) [inline R/get_bench.r]:
```suggestion
          (lubridate::year(Bench_startDate) <= year &
             is.na(Bench_endDate)))
```

Comment (adamgrimes) [reply R/get_bench.r]: This removed two seemingly erroneous rows for me. One had a start date of 2018-12-03 with no end date, the other a start date of 2020-01-09 with no end date

**Review by mark-druffel — CHANGES_REQUESTED** [review]
Overall notes below. Let me know if anything here is confusing or you disagree and we can talk through it. I don't think any of this is mission critical so let me know if you want me to put these in trello for future and approve. 

**Replace dplyr::as_data_frame**
As @adamgrimes mentioned, we should replace dplyr::as_data_frame as it's been deprecated for awhile. I tested tibble::enframe and it works, but depending on the context it should either be tibble::enframe or tibble::as_tibble. These deeply nested lists are very difficult to work with so long-term we may even want to move to more explicit code rather than relying on all coercion... I'm just not sure of the best way at this point. 

**Update Manuals for New Parameters**
The new date parameters are in the manuals. We should add them and include syntax. Dates needed to be 'YYYY-mm-dd' for get_pto which isn't necessarily obvious. Users will get a parse error if they use '/'. 

**Coerce startDate & endDate**
If startDate & endDate are not dates you get the below message. Why not automatically coerce the dates for the user? 
Error in bambooR::get_pto(user = user, password = password, startDate = startDate,  : 
  A startDate is required and must be a date!!

File: R/get_pto.r (hunk 2 above)
Comment (cfry-propeller) [reply R/get_pto.r]: I'm curious about why this is necessary if you got the same data before and after. What's motivating the change? 

I also don't know anywhere near enough about the inner workings of the tidyverse to say what's right here... @mark-druffel what's the diff between these two approaches?

File: R/get_pto.r (hunk 3 above)
Comment (cfry-propeller) [reply R/get_pto.r]: @mark-druffel  Same question here... what's the diff between dplyr::as_data_frame() and tibble::enframe() in this scenario? Which should be used?

File: R/get_bench.r (hunk 4 above)
Comment (cfry-propeller) [reply R/get_bench.r]: I think your code change is right - good catch! There's a subtlety here that I want to make sure you're aware of, though...

The way the process works today the Ops team creates a bench record with no end date when someone hits the bench. It stays open until they start a project, at which time the Ops team goes in again and sets the end date. We therefore interpret no end date to mean "still on the bench today", which is what I was trying to account for. 

So, the one with a start date of 2018-12-03 and no end is clearly just setup wrong, but the function should return it. As far as we know, this is a person who's been on the bench for the last 13 months. 

The one with a startDate of 2020-01-09 you're right about. This was in the future when you executed the code and I don't see any value in returning future bench records (though the rest of our code wouldn't be hurt by this).

Your change should filter out the latter but keep the former, which is right.

**Comment by cfry-propeller** [comment]
@mark-druffel  Apologies, I see that you already responded on @adamgrimes observation on as_data_frame vs enframe - I'll commit those changes.

@adamgrimes kindly supplied an example with the new date params, so I think that comment is covered.

startDate & endDate must be dates. I'm surprised you are able to pass in a string date with hyphens like '2019-12-31'. there must be some automatic coercion taking care of it for you somewhere along the way as I would expect that to fail when the check that you commented on occurs.

I'm not super into the idea of coercing the value for the caller. You could have situations where the passed value is successfully coerced to a date when in fact it wasn't a date, which could lead to odd situations (e.g. user passes an integer 2020 and casting as a date turns it into 2020-01-01 - not sure that's a real situation, but you see how this could be problematic). The caller should create a Date object and pass that to be explicit about what they mean. This ensures the function will behave as they request.

**Pull request merged**